snapfaas/bins/multivm/main.rs

```diff
@@ -85,57 +92,54 @@ fn main() {
+    let total_mem = matches
+                        .value_of("total memory")
+                        .unwrap()
+                        .parse::<usize>()
+                        .expect("Total memory is not a valid integer");
     manager.set_total_mem(total_mem);
 
     // create the worker pool
-    let (pool, request_sender) = new_workerpool(manager.total_mem()/128, manager_sender.clone());
+    let pool = new_workerpool(manager.total_mem()/128, sched_addr.clone(), manager_sender.clone());
     // kick off the resource manager
     let manager_handle = manager.run();
 
     // register signal handler
-    set_ctrlc_handler(request_sender.clone(), pool, manager_sender, Some(manager_handle));
-
-    // TCP gateway
-    if let Some(l) = matches.value_of("listen address") {
-        let gateway = gateway::HTTPGateway::listen(l);
+    set_ctrlc_handler(pool, sched_addr.clone(), manager_sender, Some(manager_handle));
 
-        for (request, response_tx, timestamps) in gateway {
-            // Return when a VM acquisition succeeds or fails
-            // but before a VM launches (if it is newly allocated)
-            // and execute the request.
-            request_sender.send(Message::Request((request, response_tx, timestamps))).expect("Failed to send request");
-        }
-    }
+    // hold on
+    let (_, rx) = mpsc::channel::<usize>();
+    loop { let _ = rx.recv(); }
 }
 
-fn new_workerpool(pool_size: usize, manager_sender: Sender<Message>) -> (Vec<Worker>, Sender<Message>) {
-    let (request_sender, response_receiver) = mpsc::channel();
-    let response_receiver = Arc::new(Mutex::new(response_receiver));
-
+fn new_workerpool(
+    pool_size: usize, sched_addr: String, manager_sender: Sender<Message>
+) -> Vec<Worker> {
     let mut pool = Vec::with_capacity(pool_size);
-
     for i in 0..pool_size {
         let cid = i as u32 + 100;
-        pool.push(Worker::new(response_receiver.clone(), manager_sender.clone(), request_sender.clone(), cid));
+        pool.push(Worker::new(
+            sched_addr.clone(),
+            manager_sender.clone(),
+            cid,
+        ));
     }
-
-    (pool, request_sender)
+    pool
 }
 
-fn set_ctrlc_handler(request_sender: Sender<Message>, mut pool: Vec<Worker>, manager_sender: Sender<Message>, mut manager_handle: Option<JoinHandle<()>>) {
+fn set_ctrlc_handler(
+    mut pool: Vec<Worker>, sched_addr: String,
+    manager_sender: Sender<Message>, mut manager_handle: Option<JoinHandle<()>>
+) {
     ctrlc::set_handler(move || {
-        println!("");
+        println!("ctrlc handler");
         warn!("{}", "Handling Ctrl-C. Shutting down...");
-        let pool_size = pool.len();
-        for _ in 0..pool_size {
-            request_sender.send(Message::Shutdown).expect("failed to shut down workers");
-        }
+        let mut sched = snapfaas::sched::rpc::Scheduler::new(sched_addr.clone());
+        let _ = sched.terminate_all();
```

Comment:
I don't think we want this behavior?  It doesn't really make sense that shutting down a worker machine bring down the entire system.

Reply:
Fixed